doc/specs/stdlib_linalg.md

```diff
@@ -206,3 +206,285 @@ program demo_outer_product
     !A = reshape([3., 6., 9., 4., 8., 12.], [3,2])
 end program demo_outer_product
 ```
+
+## `is_square` - Checks if a matrix is square
+
+### Status
+
+Experimental
+
+### Description
+
+Checks if a matrix is square
+
+### Syntax
+
+`d = [[stdlib_linalg(module):is_square(interface)]](A)`
+
+### Arguments
+
+`A`: Shall be a rank-2 array
+
+### Return value
+
+Returns a logical value that is true if the input matrix is square, and false otherwise.
+
+### Example
+
+```fortran
+program demo_is_square
+    use stdlib_linalg, only: is_square
+    implicit none
+    real :: A_true(2,2), A_false(3,2)
+    logical :: res
+    A_true = reshape([1., 2., 3., 4.], shape(A_true))
+    A_false = reshape([1., 2., 3., 4., 5., 6.], shape(A_false))
+    res = is_square(A_true)
+    !res = .true.
+    res = is_square(A_false)
+    !res = .false.
+end program demo_is_square
+```
+
+## `is_diagonal` - Checks if a matrix is diagonal
+
+### Status
+
+Experimental
+
+### Description
+
+Checks if a matrix is diagonal
+
+### Syntax
+
+`d = [[stdlib_linalg(module):is_diagonal(interface)]](A)`
+
+### Arguments
+
+`A`: Shall be a rank-2 array
+
+### Return value
+
+Returns a logical value that is true if the input matrix is diagonal, and false otherwise.
+Note that nonsquare matrices may be diagonal, so long as `a_ij = 0` when `i /= j`.
+
+### Example
+
+```fortran
+program demo_is_diagonal
+    use stdlib_linalg, only: is_diagonal
+    implicit none
+    real :: A_true(2,2), A_false(2,2)
+    logical :: res
+    A_true = reshape([1., 0., 0., 4.], shape(A_true))
+    A_false = reshape([1., 0., 3., 4.], shape(A_false))
+    res = is_diagonal(A_true)
+    !res = .true.
+    res = is_diagonal(A_false)
+    !res = .false.
+end program demo_is_diagonal
+```
+
+## `is_symmetric` - Checks if a matrix is symmetric
+
+### Status
+
+Experimental
+
+### Description
+
+Checks if a matrix is symmetric
+
+### Syntax
+
+`d = [[stdlib_linalg(module):is_symmetric(interface)]](A)`
+
+### Arguments
+
+`A`: Shall be a rank-2 array
+
+### Return value
+
+Returns a logical value that is true if the input matrix is symmetric, and false otherwise.
+
+### Example
+
+```fortran
+program demo_is_symmetric
+    use stdlib_linalg, only: is_symmetric
+    implicit none
+    real :: A_true(2,2), A_false(2,2)
+    logical :: res
+    A_true = reshape([1., 3., 3., 4.], shape(A_true))
+    A_false = reshape([1., 0., 3., 4.], shape(A_false))
+    res = is_symmetric(A_true)
+    !res = .true.
+    res = is_symmetric(A_false)
+    !res = .false.
+end program demo_is_symmetric
```

Comment:
```suggestion
program demo_is_symmetric
    use stdlib_linalg, only: is_symmetric
    implicit none
    real :: A(2,2), B(2,2)
    logical :: res
    A = reshape([1., 3., 3., 4.], shape(A))
    B = reshape([1., 0., 3., 4.], shape(B))
    res = is_symmetric(A) ! returns .true.
    res = is_symmetric(B) ! returns .false.
end program demo_is_symmetric
```

In light of the discussion at https://github.com/fortran-lang/stdlib/pull/539#discussion_r717399771, one more nitpick would be to revise the demonstration examples. 

@ghbrown, @milancurcic How do you feel about the clarity of my proposed patch?

Reply:
I like it!

Reply:
I don't have a strong opinion on this one, but since other people seem to like the `return` version better I'm happy to make the change.
I'll go ahead and edit the other tests as well.

Reply:
All examples should be updated, thanks for the feedback.